quinn-udp/src/windows.rs

```diff
@@ -62,30 +109,98 @@ impl UdpSocketState {
         })
     }
 
-    pub fn send(&self, socket: UdpSockRef<'_>, transmits: &[Transmit]) -> Result<usize, io::Error> {
+    pub fn send(&self, socket: UdpSockRef<'_>, transmits: &[Transmit]) -> io::Result<usize> {
         let mut sent = 0;
         for transmit in transmits {
-            match socket.0.send_to(
-                &transmit.contents,
-                &socket2::SockAddr::from(transmit.destination),
-            ) {
-                Ok(_) => {
-                    sent += 1;
+            // we cannot use [`socket2::sendmsg()`] and [`socket2::MsgHdr`] as we do not have access
+            // to the inner field which holds the WSAMSG
+            let mut ctrl_buf = cmsg::Aligned([0; CMSG_LEN]);
+            let daddr = socket2::SockAddr::from(transmit.destination);
+
+            let mut data = WinSock::WSABUF {
+                buf: transmit.contents.as_ptr() as *mut _,
+                len: transmit.contents.len() as _,
+            };
+
+            let ctrl = WinSock::WSABUF {
+                buf: ctrl_buf.0.as_mut_ptr(),
+                len: ctrl_buf.0.len() as _,
+            };
+
+            let mut wsa_msg = WinSock::WSAMSG {
+                name: daddr.as_ptr() as *mut _,
+                namelen: daddr.len(),
+                lpBuffers: &mut data,
+                Control: ctrl,
+                dwBufferCount: 1,
+                dwFlags: 0,
+            };
+
+            // Add control messages (ECN and PKTINFO)
+            let mut encoder = unsafe { cmsg::Encoder::new(&mut wsa_msg) };
+
+            if let Some(ip) = transmit.src_ip {
+                let ip = std::net::SocketAddr::new(ip, 0);
+                let ip = socket2::SockAddr::from(ip);
+                match ip.family() {
+                    WinSock::AF_INET => {
+                        let src_ip: WinSock::SOCKADDR_IN = unsafe { ptr::read(ip.as_ptr() as _) };
```

Comment:
Style nit: feels silly to have a type annotation combined with `as _`. For those cases, let's just write it as `as WinSock::SOCKADDR_IN`?